maps-compose/src/main/java/com/google/maps/android/compose/MapViewLifecycleController.kt

```diff
@@ -0,0 +1,136 @@
+package com.google.maps.android.compose
+
+import android.os.Bundle
+import android.util.Log
+import androidx.lifecycle.Lifecycle
+import androidx.lifecycle.Lifecycle.Event
+import androidx.lifecycle.LifecycleEventObserver
+import com.google.android.gms.maps.MapView
+import com.google.maps.android.compose.MapViewLifecycleController.LifecycleDirection.Down
+import com.google.maps.android.compose.MapViewLifecycleController.LifecycleDirection.Up
+
+internal class MapViewLifecycleController(
+    isMapReused: Boolean,
+    private val mapView: MapView
+) {
+    private val lcTag = run {
+        val mapViewId = mapView.tagData().debugId
+        "MVLC/$mapViewId"
+    }
+
+    private companion object {
+        val lifecycleUp = listOf(
+            Event.ON_CREATE,
+            Event.ON_START,
+            Event.ON_RESUME
+        )
+    }
+
+    private enum class LifecycleDirection {
+        Up, Down
+    }
+
+    private var previousLifecycleState = if (isMapReused)
+        Event.ON_STOP else
+        Event.ON_CREATE
+
+    var lifecycle: Lifecycle? = null
+        set(value) {
+            if (field !== value) {
+                field?.removeObserver(observer)
+                value?.addObserver(observer)
+                field = value
+            }
+        }
+
+    /**
+     * Move to the lifecycle event instead of setting it directly.
+     * */
+    private fun moveToLifecycleEvent(targetLifecycle: Event) {
+        if (targetLifecycle == previousLifecycleState) return
+
+        val lifecycleDirection = if (targetLifecycle in lifecycleUp) Up else Down
+
+        (if (lifecycleDirection == Up) "UP" else "DOWN").let { strDirection ->
+            Log.d(lcTag, "Moving $strDirection from $previousLifecycleState to $targetLifecycle.")
+        }
+
+        do {
+            val nextLifecycleState = nextLifecycleEvent(lifecycleDirection)
+            setLifecycleEvent(nextLifecycleState)
+        } while (nextLifecycleState != targetLifecycle)
+    }
+
+    private fun nextLifecycleEvent(direction: LifecycleDirection) = when (previousLifecycleState) {
+        Event.ON_CREATE -> when (direction) {
+            Up -> Event.ON_START
+            Down -> error("No lifecycle state below created.")
+        }
+
+        Event.ON_START -> when (direction) {
+            Up -> Event.ON_RESUME
+            Down -> error("No lifecycle state below started.")
+        }
+
+        Event.ON_RESUME -> when (direction) {
+            Up -> error("No lifecycle state above resumed.")
+            Down -> Event.ON_PAUSE
+        }
+
+        Event.ON_PAUSE -> when (direction) {
+            Up -> Event.ON_RESUME
+            Down -> Event.ON_STOP
+        }
+
+        Event.ON_STOP -> when (direction) {
+            Up -> Event.ON_START
+            Down -> Event.ON_DESTROY
+        }
+
+        Event.ON_DESTROY -> error("No lifecycle state above destroyed")
+        else -> error("Unsupported operation")
```

Comment:
I rewrote MapViewLifecycleController.
You were right that it could be simplified and I believe it's much simpler and more idiomatic now.
I also removed handling of changing lifecycles.

`previousLifecycleState` and `isMapReused` were removed as well.

Reply:
Still looks like there is a large amount of code dedicated to handling lifecycle things.

Would you be willing to consider an alternative approach? It holds the promise of largely getting rid of all the special cases & processing. What I'm hoping for is an approach where things fall into place without much custom coding because the code logic aligns well with its environment.

The `AndroidView()` docs have this [blurb](https://developer.android.com/reference/kotlin/androidx/compose/ui/viewinterop/package-summary#AndroidView(kotlin.Function1,androidx.compose.ui.Modifier,kotlin.Function1,kotlin.Function1,kotlin.Function1):~:text=If%20you%20need,View%20is%20attached.):

> If you need to observe whether the View is currently used in the composition hierarchy, you may observe whether it is attached via [View.addOnAttachStateChangeListener](https://developer.android.com/reference/android/view/View.html#addOnAttachStateChangeListener(android.view.View.OnAttachStateChangeListener)). The View may also observe the lifecycle of its host via [findViewTreeLifecycleOwner](https://developer.android.com/reference/kotlin/androidx/lifecycle/package-summary#findViewTreeLifecycleOwner(android.view.View)). The lifecycle returned by this function will match the [LocalLifecycleOwner](https://developer.android.com/reference/kotlin/androidx/lifecycle/compose/package-summary#LocalLifecycleOwner()). Note that the lifecycle is not set and cannot be used until the View is attached.

I needed something similar a while back. It looks kinda like this:

```kotlin
private val lifecycleObserver: LifecycleObserver =
    TODO("LifecycleObserver that merely passes events on to MapView")

view.addOnAttachStateChangeListener(
    object : View.OnAttachStateChangeListener {
        override fun onViewAttachedToWindow(v: View) {
            val lifecycle = v.findViewTreeLifecycleOwner()!!.lifecycle

            lifecycle.addObserver(lifecycleObserver)
        }

        // The detach event might be a bit unreliable: not sure if it's always
        // called (in the case where the activity is destroyed),
        // but the lifecycle & activity will be destroyed in this case anyway,
        // and all references should be released
        override fun onViewDetachedFromWindow(v: View) {
            //TODO not sure if we're already detached here,
            // then findViewTreeLifecycleOwner would likely return null;
            // would need to save owner above in this case,
            // but that's a little more messy
            val lifecycle = v.findViewTreeLifecycleOwner()!!.lifecycle

            lifecycle.removeObserver(lifecycleObserver)
        }
    }
)
```

I think this can reside in the `factory` lambda, so it's well-encapsulated.

There is a little uncertainty here about `onDestroy()`: `lifecycle.removeObserver()` may need to be called from LifecycleObserver.onDestroy() instead of the above, otherwise the observer might end up being removed before onDestroy() is called on it. However, the documentation for `View.setViewTreeLifecycleObserver()` indicates that `onDestroy()` will be called on detach if not guaranteed to reattach, so that's vague, too.

I'm not sure if `MapView` tolerates repeated `onCreate()`/`onDestroy()` cycles. In this case they'd have to be skipped in the LifecycleObserver and be called once from elsewhere. Perhaps that would be the best route here, but all this needs a bit of testing to validate actual behaviors.

Reply:
Also, not sure if you saw those, but Lifecycle and LifecycleRegistry have some sample code and functions to assist with raising/lowering state gradually.

Reply:
For starters it might work best if you gradually lower the MapView lifecycle to Destroyed from `onViewDetachedFromWindow()`, to eliminate race conditions, in addition to removing the observer there.

Then, if `MapView` cannot handle repeated onCreate/onDestroy, handle those two separately, as mentioned.

Reply:
Hi, bubenheimer.

Thanks for your patience.

I am open to whatever is simply the best!


After experimenting and trying to add/remove the `LifecycleObserver` depending on attached state I think the simplest way is to keep a `LifecycleRegistry` running during the `MapView`'s entire lifetime and to remove it when it's discarded (could be in either `onDestroy()`, `onRelease` or both).

The only(?) drawback (is it?) is that the `LifecycleObserver` has to be stored somewhere (in the tag) between compositions.
Other than that it's currently working flawlessly and after using `LifecycleRegistry` it's very little code and it's pretty easy to follow imo.

But if you have a better idea/solution in mind I'm all for it. I'll experiment some more with `LifecycleObserver` add/remove on attach/detach.
Also if you'd like I can invite you to my fork.
Additionally [I'm in the Kotlin slack](https://kotlinlang.slack.com/team/U057726PH6H) if you'd prefer to discuss more there.

Thanks for the tip on LifecycleRegistry btw, it's great!

Reply:
I made another commit on [another branch](https://github.com/philip-segerfast/android-maps-compose/tree/bubenheimer-idea) which I think is kind of what you asked for.
As far as my tests go it works fine.

Reply:
Both solutions require to store something in the tag across compositions (as far as I can see). One stores a LifecycleRegistry and the other a callback for onRelease.